test/framework/config.go

```diff
@@ -0,0 +1,325 @@
+	Type ComponentSourceType `json:"type,omitempty"`
+
+	// Replacements is a list of patterns to replace in the component YAML
+	// prior to application.
+	Replacements []ComponentReplacement `json:"replacements,omitempty"`
+}
+
+// ComponentWaiterType indicates the type of check to use to determine if the
+// installed components are ready.
+type ComponentWaiterType string
+
+const (
+	// ServiceWaiter indicates to wait until a service's condition is Available.
+	// When ComponentWaiter.Value is set to "service", the ComponentWaiter.Value
+	// should be set to the name of a Service resource.
+	ServiceWaiter ComponentWaiterType = "service"
+
+	// PodsWaiter indicates to wait until all the pods in a namespace have a
+	// condition of Ready.
+	// When ComponentWaiter.Value is set to "pods", the ComponentWaiter.Value
+	// should be set to the name of a Namespace resource.
+	PodsWaiter ComponentWaiterType = "pods"
+)
+
+// ComponentWaiter contains information to help determine whether installed
+// components are ready.
+type ComponentWaiter struct {
+	// Value varies depending on the specified Type.
+	// Please see the documentation for the different WaiterType constants to
+	// understand the valid values for this field.
+	Value string `json:"value"`
+
+	// Type describes the type of check to perform.
+	//
+	// Defaults to "pods".
+	Type ComponentWaiterType `json:"type,omitempty"`
+}
+
+// ComponentReplacement is used to replace some of the generated YAML prior
+// to application.
+type ComponentReplacement struct {
+	// Old is the pattern to replace.
+	// A regular expression may be used.
+	Old string `json:"old"`
+	// New is the string used to replace the old pattern.
+	// An empty string is valid.
+	New string `json:"new,omitempty"`
+}
+
+// ComponentConfig describes a component required by the e2e test environment.
+type ComponentConfig struct {
+	// Name is the name of the component.
+	// This field is primarily used for logging.
+	Name string `json:"name"`
+
+	// Sources is an optional list of component YAML to apply to the management
+	// cluster.
+	// This field may be omitted when wanting only to block progress via one or
+	// more Waiters.
+	Sources []ComponentSource `json:"sources,omitempty"`
+
+	// Waiters is an optional list of checks to perform in order to determine
+	// whether or not the installed components are ready.
+	Waiters []ComponentWaiter `json:"waiters,omitempty"`
+}
+
+// Config is the input used to configure the e2e test environment.
+type Config struct {
+	// Name is the name of the Kind management cluster.
+	// Defaults to DefaultManagementClusterName.
+	ManagementClusterName string `json:"managementClusterName,omitempty"`
+
+	// KubernetesVersion is the version of Kubernetes to deploy when testing.
+	// Defaults to DefaultKubernetesVersion.
+	KubernetesVersion string `json:"kubernetesVersion,omitempty"`
+
+	// Images is a list of container images to load into the Kind cluster.
+	Images []string `json:"images,omitempty"`
```

Comment:
Is there anything we are doing to ensure that the Images listed match up with the images referenced in the component configs? Should we just avoid listing images separately and use what is in the component configs instead? I believe @fabriziopandini is doing something along these lines in the `cluterctl` work.

Reply:
Hi @detiber,

This is by design. The images from the YAML may not be available locally. This lets you manually load any images you *did* build locally. There's a clear separation of images and the component YAML, and that's on purpose.